examples/tests/timer_overflow/main.c

```diff
@@ -0,0 +1,80 @@
+#include "internal/alarm.h"
+#include "timer.h"
+#include "tock.h"
+
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+
+/*
+ * This test registers a "long running" timer and verifies that the
+ * timer fires at about the time when it's expected to.
+ *
+ * NOTE: Modify the "MAX_TICKS" macro in "libtock/alarm_timer.c" to
+ * test the implementation of timer_in without waiting for
+ * the alarm to overflow after 2^32 ticks (which is about 36 hours
+ * with a 32768Hz oscillator). Example: if you want to simulate the
+ * alarm overflowing after a second on a 32768Hz oscillator you can set
+ * "MAX_TICKS" to 32768. You can get the frequence of the alarm by calling
+ * alarm_internal_frequency.
+ */
+
+#define TIMER_LEN_MS 2000
+
+// helper struct to send data to callback
+typedef struct testing_ud {
+  uint32_t timer_start_ticks;
+  uint32_t timer_len_ms;
+  uint32_t expected_timer_end_ticks;
+} testing_ud_t;
+
+static bool fired = false;
+
+static void timer_fire_callback(__attribute__ ((unused)) int   unused0,
+                                __attribute__ ((unused)) int   unused1,
+                                __attribute__ ((unused)) int   unused2,
+                                __attribute__ ((unused)) void* unused3) {
+  fired = true;
+}
+
+// copied this function over from alarm_timer.c. we could expose it in a header
+// to avoid duplicated code.
+inline static uint32_t ms_to_ticks(uint32_t ms) {
+  uint32_t frequency;
+  alarm_internal_frequency(&frequency);
+
+  uint32_t millihertz      = frequency / 1000; // ticks per millisecond
+  uint32_t seconds         = ms / 1000;
+  uint32_t leftover_millis = ms % 1000;
+  return (seconds * frequency) + (leftover_millis * millihertz);
+}
+
+int main(int argc, char** argv)
+{
+  // need to allocate memory for a timer before giving a pointer to `timer_in`
+  tock_timer_t* timer = malloc(sizeof(tock_timer_t));
+
+  uint32_t timer_start_ticks;
+  alarm_internal_read(&timer_start_ticks);
+  testing_ud_t* ud = malloc(sizeof(testing_ud_t));
+  ud->timer_len_ms             = TIMER_LEN_MS;
+  ud->timer_start_ticks        = timer_start_ticks;
+  ud->expected_timer_end_ticks = timer_start_ticks + ms_to_ticks(TIMER_LEN_MS);
```

Comment:
Does the current design require this information be passed to `timer_in`? Ideally, `timer_in` would abstract needing to allocate this struct and just be passed a time in ms and a timer object.

Reply:
You can ignore the `ud` (user data) part of the timer by passing in a nullptr. `timer_in` does not use this at all - it is mechanism to pass data to your callback. As `ud` is effectively an optional parameter, I do not see any reason to add another function stub that omits `ud`.
Using `ud` in this test means that we also check that we do not corrupt this field.

Reply:
As for the usage of `ud` here, these are values that we wanted to send to the callback for our testing setup. Originally we were printing out these values in the callback but we moved the prints to the end of the file after the `yield_for` so these don't need to be passed as user data.

Reply:
To avoid any confusion we've removed this `testing_ud_t` struct and are passing in a single integer as user data. There's also a comment that explains how the user_data parameter is completely optional and that we are only using it for printing out values in the callback. The comment mentions that they can pass NULL as `ud` if they don't need to pass any user data.